### PR TITLE
Fixed issue where pause throws an error

### LINKED ISSUE #6935: Cloud Storage pause in uploadBytesResumable not work with vue.js
* Operating System version: Ubuntu 22.04.1 LTS
  * Browser version: google chrome 108.0.5359.124 (Official Build) (64-bit)
  * Firebase SDK version: 9.15.0
  * Firebase Product: Cloud Storage

### the problem
pause method in uploadBytesResumable not work with vue.js
when use pause
it says : 

> Firebase Storage: An unknown error occurred, please check the error payload for server response. (storage/unknown)

but with firebase@9.7.0 there is no problem
**Live DEMO** : [https://stackblitz.com/edit/vue-qtohv9](https://stackblitz.com/edit/vue-qtohv9)

### PR DESCRIPTION
**Pull request by maneesht**
Fixes #6935

### CODE REVIEW

**Review by tonyjhuang — APPROVED** [review]

File: packages/storage/test/unit/task.test.ts [hunk 1]
```diff
@@ -341,7 +341,137 @@ describe('Firebase Storage > Upload Task', () => {
...
+    const storageService = storageServiceWithHandler(
+      fakeServerHandler(),
+      shouldRespondCallback
+    );
+    const task = new UploadTask(
+      new Reference(storageService, testLocation),
+      blob
+    );
+
+    // eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (maneesht) [inline packages/storage/test/unit/task.test.ts:370]: Double check linting of this

File: packages/storage/test/unit/task.test.ts [hunk 2]
```diff
@@ -341,7 +341,137 @@ describe('Firebase Storage > Upload Task', () => {
...
+    const promise = new Promise<void>((innerResolve, innerReject) => {
+      resolve = innerResolve;
+      reject = innerReject;
+    });
 
+    // Assert functions throw Errors, which means if we're not in the right stack
+    // the error might not get reported properly. This function wraps existing
+    // assert functions, returning a new function that calls reject with any
+    // caught errors. This should make sure errors are reported properly.
+    function promiseAssertWrapper<T>(func: T): T {
```
Comment (maneesht) [inline packages/storage/test/unit/task.test.ts:381]: We can move this into a helper

File: packages/storage/test/unit/task.test.ts [hunk 3]
```diff
@@ -341,7 +341,137 @@ describe('Firebase Storage > Upload Task', () => {
     });
 
     task.pause();
+    return promise;
+  }
+
+  // This is to test to make sure that when you pause in the middle of a request that you do not get an error
+  async function runProgressPauseTest(blob: FbsBlob): Promise<void> {
```
Comment (maneesht) [inline packages/storage/test/unit/task.test.ts:348]: Explain more here

File: packages/storage/test/unit/task.test.ts [hunk 4]
```diff
@@ -341,7 +341,137 @@ describe('Firebase Storage > Upload Task', () => {
...
+      }
+
+      fixedAssertTrue(increasing);
+      fixedAssertTrue(allTotalsTheSame);
+      fixedAssertTrue(lastIsAll);
+      resolve(null);
+    });
+    await pausedDeferred.promise;
+    // Need to wait until the state has settled [i.e. from pausing to pause] and allow for any potential errors to propagate.
+    await new Promise(resolve => setTimeout(resolve, 1000));
```
Comment (maneesht) [inline packages/storage/test/unit/task.test.ts:473]: Add another promise to wait until pausing has settled

File: packages/storage/test/unit/task.test.ts [hunk 5]
```diff
@@ -341,7 +341,140 @@ describe('Firebase Storage > Upload Task', () => {
...
+    const fixedAssertFalse = promiseAssertWrapper(assert.isFalse);
+    const fixedAssertTrue = promiseAssertWrapper(assert.isTrue);
+    const fixedAssertFail = promiseAssertWrapper(assert.fail);
+
+    const events: string[] = [];
+    const progress: number[][] = [];
+    // Promise for when we are finally in the pause state
+    const pausedStateCompleted = new Deferred();
+    let complete = 0;
+    function addCallbacks(task: UploadTask): void {
```
Comment (tonyjhuang) [inline packages/storage/test/unit/task.test.ts:403]: do you mind documenting this function as well

**Review by tonyjhuang — APPROVED** [review]
one small comment

**Review by hsubox76 — APPROVED** [review]

**Pull request merged**